pdns/slavecommunicator.cc

```diff
@@ -615,7 +615,17 @@ void CommunicatorClass::suck(const DNSName &domain, const ComboAddress& remote)
     }
   }
   catch(ResolverException &re) {
-    g_log<<Logger::Error<<"Unable to AXFR zone '"<<domain<<"' from remote '"<<remote<<"' (resolver): "<<re.reason<<endl;
+    // The AXFR probably failed due to a problem on the master server. If SOA-checks against this master
+    // still succeed, we would constantly try to AXFR the zone. To avoid this, we add the zone to the list of
+    // failed slave-checks. This will suspend slave-checks (and subsequent AXFR) for this zone for some time.
+    uint64_t newCount = 1;
+    time_t now = time(0);
+    const auto failedEntry = d_failedSlaveRefresh.find(domain);
```

Comment:
We need to acquire `d_lock` before accessing `d_failedSlaveRefresh`.

Reply:
Can you point me to an example how to use d_lock?

Reply:
Sure, sorry! You need to acquire the lock by doing something like:
```
Lock l(&d_lock);
```
It will create a `l` object that will hold the lock as long as it exists, and will release it as soon as it is destroyed. In this case the easiest way would be to create it right after entering the `catch` block you are in, and then just let it be destroyed when you exit this block.